File: 20210523LangReview/Python/08-interface/interface.cs
/*

this module is to demonstrate interface concept in C#

@ZL, 20210526

[ conclusion ]
===
- C# does support interface
===

*/

using System;

namespace TS
{

    interface Pack
    {
        virtual string start();
    }

    class Carton: Pack
    {
        private string m_name = "noname";
        private int m_age = 0;

        public Carton(string name, int age)
        {
            m_name = name;
            m_age = age;
        }

        override string start()
        {
            return "class Carton says hello";
        } 
    }

    class TS
    {
        static void Main(string[] args)
        {
            const Carton c = new Carton("TS", 30);
            System.Console.WriteLine(c.start());
            Console.ReadLine();
        }
    }
}

File: 20210523LangReview/Python/08-interface/interface.cpp
/*

this module is to demonstrate interface concept in C++

@ZL, 20210526

[ conclusion ]
===
- C++ does not support interface
- but it does have a walkaround, using class
===

*/

#include <string>
#include <iostream>

struct Pack {
    virtual std::string start(void) = 0;
};

class Carton: public Pack
{
    private:
        std::string m_name;
        int m_age;

    public:
        Carton(const std::string& name, int age)
        : m_name(name), m_age(age)
        {}

        std::string start(void) override
        {
            return "class Carton says hello";
        }
};

int main()
{
    Carton* c = new Carton("TS", 30);
    std::cout << c->start() << std::endl;

    delete c;

    return 0;
}
